File: 0859-buddy-strings/0859-buddy-strings.cpp
class Solution {
public:
    bool buddyStrings(string s, string goal) {
        if(s.length() != goal.length()) return false;
        if(s == goal) {
            vector<int> frequency(26, 0);
            for(char ch : s) {
                frequency[ch - 'a'] += 1;
                if(frequency[ch - 'a'] == 2) {
                    return true;
                }
            }
            return false;
        }
        int n = s.length();
        vector<int> indices;
        for(int i = 0; i < n; i++) {
            if(s[i] != goal[i]) {
                indices.push_back(i);
            }
        }
        if(indices.size() != 2) return false;
        int idx1 = indices[0];
        int idx2 = indices[1];
        return s[idx1] == goal[idx2] && s[idx2] == goal[idx1];
    }
};

File: 0859-buddy-strings/0859-buddy-strings.py
class Solution(object):
    def buddyStrings(self, s, goal):
        if len(s) != len(goal):
            return False
        
        if s == goal:
            frequency = [0] * 26
            for ch in s:
                frequency[ord(ch) - ord('a')] += 1
                if frequency[ord(ch) - ord('a')] == 2:
                    return True
            return False
        
        n = len(s)
        indices = []
        for i in range(n):
            if s[i] != goal[i]:
                indices.append(i)
        
        if len(indices) != 2:
            return False
        
        idx1, idx2 = indices[0], indices[1]
        return s[idx1] == goal[idx2] and s[idx2] == goal[idx1]
